File: iOS/NatReader/Bridge.m
//
//  Bridge.m
//  NatReader
//
//  Created by Yusuf Olokoba on 9/23/19.
//  Copyright © 2020 Yusuf Olokoba. All rights reserved.
//

#import "NatReader.h"
#import "NRMediaReader.h"

void NRDisposeReader (void* readerPtr) {
    id<NRMediaReader> reader = (__bridge_transfer id<NRMediaReader>)readerPtr;
    reader = nil;
}

void NRMediaURI (void* readerPtr, char* dstString) {
    id<NRMediaReader> reader = (__bridge id<NRMediaReader>)readerPtr;
    strcpy(dstString, reader.uri.absoluteString.UTF8String);
}

float NRMediaDuration (void* readerPtr) {
    id<NRMediaReader> reader = (__bridge id<NRMediaReader>)readerPtr;
    return reader.duration;
}

void* NRCreateMP4FrameReader (const char* url) {
    NSURL* uri = [NSURL fileURLWithPath:[NSString stringWithUTF8String:url]];
    NRMP4FrameReader* reader = [NRMP4FrameReader.alloc initWithURI:uri];
    return (__bridge_retained void*)reader;
}

void NRFrameSize (void* frameReaderPtr, int32_t* width, int32_t* height) {
    id<NRFrameReader> reader = (__bridge id<NRFrameReader>)frameReaderPtr;
    *width = reader.frameSize.width;
    *height = reader.frameSize.height;
}

float NRFrameRate (void* frameReaderPtr) {
    id<NRFrameReader> reader = (__bridge id<NRFrameReader>)frameReaderPtr;
    return reader.frameRate;
}

void* NRCreateEnumerator (void* readerPtr, float startTime, float duration) {
    id<NRMediaReader> reader = (__bridge id<NRMediaReader>)readerPtr;
    CMTime startCMTime = CMTimeMakeWithSeconds(startTime, NSEC_PER_SEC);
    CMTime durationCMTime = CMTimeMakeWithSeconds(duration, NSEC_PER_SEC);
    CMTimeRange timeRange = CMTimeRangeMake(startCMTime, durationCMTime);
    id<NRMediaEnumerator> enumerator = [reader createEnumeratorForTimeRange:timeRange];
    return (__bridge_retained void*)enumerator;
}

void NRDisposeEnumerator (void* enumeratorPtr) {
    id<NRMediaEnumerator> enumerator = (__bridge_transfer id<NRMediaEnumerator>)enumeratorPtr;
    [enumerator dispose];
}

void NRCopyNextFrame (void* enumeratorPtr, void* buffer, int32_t* outBufferSize, int64_t* outTimestamp) {
    id<NRMediaEnumerator> enumerator = (__bridge id<NRMediaEnumerator>)enumeratorPtr;
    [enumerator copyNextFrame:buffer withSize:outBufferSize andTimestamp:outTimestamp];
}

File: Windows/Bridge.cpp
//
// 	Bridge.cpp
// 	NatReader
//
//  Created by Yusuf Olokoba on 11/15/19.
//  Copyright (c) 2020 Yusuf Olokoba. All rights reserved.
//

#include "pch.hpp"
#include "NatReader.h"
#include "IMediaReader.hpp"

void* NRCreateMP4FrameReader (const char* recordingPath, float startTime, float duration) {
	return static_cast<void*>(new MP4FrameReader(recordingPath, startTime, duration));
}

void NRMediaURI (void* readerPtr, char* dstString) {
	IMediaReader* reader = static_cast<IMediaReader*>(readerPtr);
	strcpy(dstString, reader->uri);
}

float NRMediaDuration (void* readerPtr) {
	IMediaReader* reader = static_cast<IMediaReader*>(readerPtr);
	return reader->Duration();
}

void NRCopyNextFrame (void* readerPtr, void* dstBuffer, int32_t* byteSize, int64_t* timestamp) {
	IMediaReader* reader = static_cast<IMediaReader*>(readerPtr);
	reader->CopyNextFrame(dstBuffer, byteSize, timestamp);
}

void NRReset (void* readerPtr) {
	IMediaReader* reader = static_cast<IMediaReader*>(readerPtr);
	reader->Reset();
}

void NRDispose (void* readerPtr) {
	IMediaReader* reader = static_cast<IMediaReader*>(readerPtr);
	delete reader;
}

void NRFrameSize (void* readerPtr, int32_t* width, int32_t* height) {
	MP4FrameReader* reader = static_cast<MP4FrameReader*>(readerPtr);
	*width = reader->FrameWidth();
	*height = reader->FrameHeight();
}

float NRFrameRate (void* readerPtr) {
	MP4FrameReader* reader = static_cast<MP4FrameReader*>(readerPtr);
	return reader->FrameRate();
}
